parlai/mturk/webapp/dev/app.jsx

```diff
@@ -1582,6 +1607,283 @@ class AssignmentPanel extends React.Component {
+      'approved': 0,
+      'rejected': 0,
+      'remain': assignments_by_worker[current_worker].assigns.length
+    }
+    this.setState({
+      current_worker: current_worker,
+      workers_remaining: workers_remaining,
+      assignments_remaining: assignments_remaining,
+      assignments_by_worker: assignments_by_worker,
+      current_worker_stats: current_worker_stats,
+    })
+  }
+
+  fetchRunData() {
+    fetch('/runs/' + this.props.run_id)
+      .then(res => res.json())
+      .then(
+        (result) => {
+          this.parseRawRun(result);
+          this.setState({
+            assignments_loading: false,
+            run_data: result
+          });
+        },
+        (error) => {
+          this.setState({
+            assignments_loading: false,
+            error: true
+          });
+        }
+      )
+  }
+
+  componentDidMount() {
+    this.setState({assignments_loading: true});
+    this.fetchRunData();
+  }
+
+  handleReview(review) {
+    // Update worker stats to reflect the type of review, then move to the
+    // next assignment
+    let worker_stats = this.state.current_worker_stats;
+    let assignments_remaining = this.state.assignments_remaining;
+    if (review == 'approve') {
+      worker_stats.approved += 1;
+      worker_stats.remain -= 1;
+      assignments_remaining -= 1;
+    } else if (review == 'reject') {
+      worker_stats.rejected += 1;
+      worker_stats.remain -= 1;
+      assignments_remaining -= 1;
+    } else if (review == 'reverse') {
+      worker_stats.approved += 1;
+      worker_stats.rejected -= 1;
+    }
+    this.setState({
+      current_worker_stats: worker_stats,
+      assignments_remaining: assignments_remaining,
+    });
+    this.nextAssignment();
+  }
+
+  nextAssignment() {
+    // Load the next valid assignment, selecting a new current worker if
+    // the current worker has no more assignments to review
+    let worker_stats = this.state.current_worker_stats;
+    let current_worker = this.state.current_worker;
+    if (worker_stats.remain == 0) {
+      current_worker = this.nextWorker();
+    }
+
+    let assignments = this.state.assignments_by_worker[current_worker].assigns;
+    let current_assignment = assignments.shift();
+    if (current_assignment) {
+      current_assignment = current_assignment.assignment_id;
+    }
+    this.setState({
+      current_assignment: current_assignment,
+      assignments_by_worker: this.state.assignments_by_worker,
+    });
+  }
+
+  nextWorker() {
+    // Move to the next worker in the remaining workers list
+    let current_worker = this.state.workers_remaining.shift();
+    let current_worker_stats = {
+      'approved': 0,
+      'rejected': 0,
+      'remain': this.state.assignments_by_worker[current_worker].assigns.length
+    }
+    this.setState({
+      current_worker: current_worker,
+      current_worker_stats: current_worker_stats,
+      workers_remaining: this.state.workers_remaining,
```

Comment:
why is it necessary to update the state with something that's already in the state?